lib/test-utils/src/Test/QuickCheck/Extra.hs

```diff
@@ -38,6 +40,44 @@ import Test.QuickCheck
 reasonablySized :: Gen a -> Gen a
 reasonablySized = scale (ceiling . sqrt @Double . fromIntegral)
 
+-- | Resizes a generator by taking the nth root of the size parameter.
+--
+-- This combinator can restore size linearity to generators composed of 'n'
+-- independent generators in the case that each generator generates values
+-- from a range that depends on the size parameter.
+--
+-- Example:
+--
+-- Suppose that we have a single generator composed of **three** independent
+-- generators, where each generator depends on the size parameter.
+--
+-- If the current value of the size parameter is 1000, then to generate a range
+-- of up to 1000 different composite values, we can resize each individual
+-- generator so that it generates up to 10 different values:
+--
+-- >>> genComposite = Composite
+-- >>>     <$> scaleToRoot 3 genA
+-- >>>     <*> scaleToRoot 3 genB
+-- >>>     <*> scaleToRoot 3 genC
+--
+scaleToRoot :: Int -> Gen a -> Gen a
```

Comment:
This is a nice idea which might be needed _very occasionally_, but in general, I don't believe we need to fret too much over the `size` parameter. It's not meant to be so precise - for example, `instance Arbitrary (a,b,c,d,e)` in QuickCheck doesn't do any such scaling. From the developer perspective, it's fairly obvious that if you have a big ADT then its total size and range will be bigger than that of a small ADT.